### PR TITLE
refactor: casl factory - datablocks

### PR DESCRIPTION
**Pull request by HayenNico**
## Description
Subsection of PR #2748 for datablocks. PR depends on #2759 to be merged first.

This unifies the datablockEndpointAccess and datablockInstanceAccess functions in CaslAbilityFactory, removes instance-level Action elements and adjusts the affected controller to accommodate the change. The datablock-specific auth code is extracted into a separate module.

## Changes:

* Replace CaslAbilityFactory.datablockInstanceAccess and CaslAbilityFactory.datablockEndpointAccess with one function CaslAbilityFactory.datablockAccess
* Code for CaslAbilityFactory.datablockAccessis factored out into new module DatablockAbility
* Remove all instance-level attachment Action elements, rename endpoint-level actions to cover both
* Adjust endpoint and instance auth logic in attachment controller

## Tests included

- [x] Included for each change/fix?
- [x] Passing? 

## Documentation
- [x] swagger documentation updated (required for API changes)
- [x] official documentation updated

## Summary by Sourcery

Unify datablock authorization into a single CASL ability and update the datablocks controller to use consolidated actions.

Enhancements:
- Extract datablock authorization rules into a dedicated DatablockAbility service and integrate it with CaslAbilityFactory via a unified datablockAccess method.
- Simplify datablock CASL actions by replacing separate endpoint/instance/admin variants with a single set of CRUD actions.
- Adjust datablocks controller authorization checks and policies to rely on the new unified datablock ability and actions.

### CODE REVIEW

File: src/casl/abilities/datablocks.ability.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    const ifAccess = { accessGroups: { $in: user.currentGroups } };
+
+    /**
+     * Authenticated user
+     */
+    can(Action.DatablockRead, Datablock, ifOwner);
+    can(Action.DatablockRead, Datablock, ifAccess);
+    can(Action.DatablockRead, Datablock, ifPublished);
+
+    can(Action.DatablockUpdate, Datablock, ifOwner);
```
[the remark this answers is not in the document]
Comment (HayenNico) [reply src/casl/abilities/datablocks.ability.ts:58]: I interpreted the previous code as an error; for one, the old check was for accessGroups only, so the ownerGroup was not permitted to update unless it was also included in accessGroups.
Additionally, the comment on the old check states "can update if in ownerGroup"

File: src/casl/abilities/datablocks.ability.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    if (
+      user.currentGroups.some(
+        (g) =>
+          this.accessGroups?.createDatasetPrivileged.includes(g) ||
+          this.accessGroups?.createDatasetWithPid.includes(g) ||
+          this.accessGroups?.createDataset.includes(g),
```
[the remark this answers is not in the document]
Comment (HayenNico) [reply src/casl/abilities/datablocks.ability.ts]: Non-issue; accessGroups comes from configuration and is always either an empty object or has all fields

File: src/casl/abilities/datablocks.ability.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  buildAbility(user: JWTUser): MongoAbility<PossibleAbilities, Conditions> {
+    const { can, build } = new AbilityBuilder(
+      createMongoAbility<PossibleAbilities, Conditions>,
+    );
+    const ifPublished = { isPublished: true };
+
+    /**
+     * Unauthenticated user
+     */
+    can(Action.DatablockRead, Datablock, ifPublished);
```
[the remark this answers is not in the document]
Comment (HayenNico) [reply src/casl/abilities/datablocks.ability.ts:39]: This was intentionally added, since the old behavior is inconsistent with every other collection. Records with `isPublished: true` should be accessible even to unauthenticated users

**Review by nitrosx — APPROVED** [review]
Please check if the AI reviews contains any applicable points

**Comment by nitrosx** [comment]
# Code Review: refactor-casl-factory-datablocks Branch

## Assessment

### Summary of Changes

This branch refactors the CASL (Authorization) layer for datablocks by extracting datablock-specific authorization logic into a dedicated service class. The changes affect 5 files:

| File | Change Type | Lines Changed |
|------|-------------|---------------|
| `src/casl/abilities/datablocks.ability.ts` | New file | +100 |
| `src/casl/action.enum.ts` | Modified | -16 (net) |
| `src/casl/casl-ability.factory.ts` | Modified | -87 |
| `src/casl/casl.module.ts` | Modified | +2 |
| `src/datablocks/datablocks.controller.ts` | Modified | -34 (net) |

**Total: 201 insertions(+), 198 deletions(-)**

### Key Changes

1. **New Dedicated Ability Class** (`datablocks.ability.ts`)
   - Created `DatablockAbility` as an injectable service
   - Encapsulates all datablock authorization logic previously scattered in `casl-ability.factory.ts`
   - Single `buildAbility()` method handles all datablock permissions

2. **Simplified Action Enum** (`action.enum.ts`)
   - **Removed** 12 old datablock actions:
     - `DatablockCreateEndpoint`, `DatablockReadEndpoint`, `DatablockUpdateEndpoint`, `DatablockDeleteEndpoint`
     - `DatablockCreateInstance`, `DatablockReadInstance`, `DatablockUpdateInstance`
     - `DatablockReadAny`, `DatablockUpdateAny`, `DatablockDeleteAny`
   - **Added** 4 new simplified actions:
     - `DatablockCreate`, `DatablockRead`, `DatablockUpdate`, `DatablockDelete`

3. **Factory Refactoring** (`casl-ability.factory.ts`)
   - Removed `datablockEndpointAccess()` method (~30 lines)
   - Removed `datablockInstanceAccess()` method (~57 lines)
   - Added delegation method `datablockAccess()` that calls `DatablockAbility.buildAbility()`
   - Removed `Datablock` import (no longer needed directly)

4. **Module Updates** (`casl.module.ts`)
   - Added `DatablockAbility` to providers array
   - Now properly injects `DatablockAbility` into `CaslAbilityFactory`

5. **Controller Updates** (`datablocks.controller.ts`)
   - Updated all policy checks to use new action names (e.g., `DatablockCreate` instead of `DatablockCreateEndpoint`)
   - Simplified `checkPermission()` method to use new ability system
   - Removed `UnauthorizedException` import (now uses `ForbiddenException` consistently)
   - Updated `findById()` to use `checkPermission()` instead of manual ability checks
   - Added proper access control filter application in `findAll()` and `count()`

### Architecture Improvement

The refactoring follows a **Separation of Concerns** principle:

**Before:**
```
CaslAbilityFactory
├── datablockEndpointAccess() - endpoint-level permissions
├── datablockInstanceAccess() - instance-level permissions
└── 15+ other resource access methods
```

**After:**
```
CaslAbilityFactory
├── datablockAccess() - delegates to DatablockAbility
└── other resource access methods

DatablockAbility (new)
└── buildAbility() - all datablock permissions
```

This pattern:
- ✅ Reduces factory class size and complexity
- ✅ Improves maintainability by isolating datablock logic
- ✅ Makes permissions more consistent and easier to test
- ✅ Follows Single Responsibility Principle

### Security Model Consistency

The new implementation maintains the same security model:

| Permission Type | Old Implementation | New Implementation | Equivalent |
|----------------|-------------------|-------------------|-----------|
| Public read | `DatablockReadInstance` with `isPublished: true` | `DatablockRead` with `ifPublished` | ✅ Yes |
| Owner read | `DatablockReadInstance` with `ownerGroup: {$in: user.currentGroups}` | `DatablockRead` with `ifOwner` | ✅ Yes |
| Access group read | `DatablockReadInstance` with `accessGroups: {$in: user.currentGroups}` | `DatablockRead` with `ifAccess` | ✅ Yes |
| Owner update | `DatablockUpdateInstance` with `ownerGroup: {$in: user.currentGroups}` | `DatablockUpdate` with `ifOwner` | ✅ Yes |
| Privileged create | Checked in `datablockInstanceAccess()` | Checked in `buildAbility()` | ✅ Yes |
| Admin access | `DatablockReadAny`, `DatablockUpdateAny` | `AccessAny` + all actions | ✅ Yes |
| Delete access | `DatablockDeleteAny` | `DatablockDelete` | ✅ Yes |

The CASL library uses **OR** semantics for multiple `can()` calls, which is preserved.

## Improvement Needed

### 1. Redundant Permission Declarations in `DatablockAbility`

**Issue:** The `buildAbility()` method contains redundant `can()` statements.

**Current Code (lines 53-55):**
```typescript
can(Action.DatablockRead, Datablock, ifOwner);
can(Action.DatablockRead, Datablock, ifAccess);
can(Action.DatablockRead, Datablock, ifPublished);
```

The third line is redundant because `ifPublished` is already declared for unauthenticated users (line 38), and CASL permissions are cumulative. However, this doesn't cause a security issue—it's just unnecessary duplication.

**Current Code (lines 81-83):**
```typescript
can(Action.DatablockCreate, Datablock);
can(Action.DatablockRead, Datablock);
can(Action.DatablockUpdate, Datablock);
```

These are already granted in the privileged group check (lines 71-72). The `AccessAny` grant (line 79) already allows all actions. Lines 81-83 are redundant.

**Recommended Fix:** Remove redundant `can()` statements after `AccessAny` is granted, as it already implies all permissions.

### 2. Inconsistent Action Naming Convention

**Issue:** The new datablock actions don't follow the same pattern as other resources.

**Comparison:**
- **Datasets:** `DatasetCreate`, `DatasetRead`, `DatasetUpdate`, `DatasetDelete` + instance actions
- **Origdatablocks:** `OrigdatablockCreate`, `OrigdatablockRead`, `OrigdatablockUpdate`, `OrigdatablockDelete` + instance actions
- **Attachments:** `AttachmentCreateEndpoint`, `AttachmentReadEndpoint`, etc. (endpoint actions) + `AttachmentCreateInstance`, etc. (instance actions)
- **Datablocks (new):** `DatablockCreate`, `DatablockRead`, `DatablockUpdate`, `DatablockDelete`

**Observation:** Other resources have both endpoint-level and instance-level actions. The new datablock actions are simpler, which is good, but the controller still uses `@CheckPolicies` decorators that expect endpoint-level checks.

**Current Pattern in Controller:**
```typescript
@CheckPolicies("datablocks", (ability: AppAbility) =>
  ability.can(Action.DatablockRead, Datablock),
)
```

This works but may be conceptually inconsistent. The `DatablockRead` action is now used for both endpoint access and instance access.

**Recommended:** This is actually an improvement over the previous overly complex system. The simplified naming is more intuitive. No change needed, but documentation would help.

### 3. Missing Instance-Level Conditions for Delete

**Issue:** The delete permission check in the controller doesn't verify ownership or access group membership before the action.

**Current Code (lines 297-305):**
```typescript
const datablockInstance =
  this.generateDatablockInstanceForPermissions(datablock);

const ability = this.caslAbilityFactory.datablockAccess(user);
const canDelete = ability.can(Action.DatablockDelete, datablockInstance);

if (!canDelete) {
  throw new ForbiddenException("Unauthorized to delete this datablock");
}
```

**Analysis:** Looking at `DatablockAbility.buildAbility()`, `DatablockDelete` is only granted to users in DELETE_GROUPS (line 86-92). There's no condition checking if the user owns the datablock or belongs to its access groups.

**Old Implementation:** Had `DatablockDeleteAny` which also didn't check ownership.

**Verdict:** This is **not a regression**—the old code had the same behavior. However, it may be a security consideration: users in DELETE_GROUPS can delete ANY datablock, regardless of ownership. This appears to be intentional based on the group naming (DELETE_GROUPS).

**Recommendation:** If this is intentional, add a comment documenting that DELETE_GROUPS have unrestricted delete access. If not, add ownership/access group conditions to delete permission.

### 4. Inconsistent Error Handling in Controller

**Issue:** Mixed use of `ForbiddenException` and error messages.

**Example (line 304):**
```typescript
throw new ForbiddenException("Unauthorized to delete this datablock");
```

**Example (line 76):**
```typescript
throw new ForbiddenException();
```

**Recommendation:** Standardize error messages. Either:
- Always use `ForbiddenException()` without a message, or
- Always include a descriptive message

The current approach is acceptable but inconsistent. For production code, descriptive messages help with debugging.

### 5. Potential Null Safety Issue

**Issue:** In `DatablockAbility`, `accessGroups` is optional but used without null checks in some places.

**Current Code (lines 60-64):**
```typescript
user.currentGroups.some(
  (g) =>
    this.accessGroups?.createDatasetPrivileged.includes(g) ||
    this.accessGroups?.createDatasetWithPid.includes(g) ||
    this.accessGroups?.createDataset.includes(g),
)
```

**Analysis:** The optional chaining (`?.`) is correctly used here. However, if `this.accessGroups` is undefined, all conditions will evaluate to false, which means privileged users won't get their permissions. This is a **critical bug**.

**Root Cause:** In the constructor (lines 21-24):
```typescript
constructor(private configService: ConfigService) {
  this.accessGroups =
    this.configService.get<AccessGroupsType>("accessGroups");
}
```

If `accessGroups` configuration is missing, `this.accessGroups` will be `undefined`.

**Impact:** 
- Users in CREATE_DATASET groups won't get create/update permissions
- Users in ADMIN groups won't get admin permissions
- Users in DELETE groups won't get delete permissions

**Recommended Fix:**
```typescript
constructor(private configService: ConfigService) {
  this.accessGroups =
    this.configService.get<AccessGroupsType>("accessGroups") ?? {}
      as AccessGroupsType;
}
```

Or better, make it required:
```typescript
constructor(private configService: ConfigService) {
  this.accessGroups =
    this.configService.getOrThrow<AccessGroupsType>("accessGroups");
}
```

### 6. Missing Type Safety in Access Control

**Issue:** The `addAccessControlFilters` method in the controller uses string-based property access.

**Current Code (lines 318-347):**
```typescript
private addAccessControlFilters(
  datablockFilter: IFilters<DatablockDocument>,
  user: JWTUser,
): IFilters<DatablockDocument> {
  datablockFilter.where = datablockFilter.where ?? {};

  if (!user) {
    if (datablockFilter.where["$and"]) {
      datablockFilter.where["$and"].push({ isPublished: true });
    } else {
      datablockFilter.where["$and"] = [{ isPublished: true }];
    }
  } else {
    const accessControlFilters = {
      $or: [
        { ownerGroup: { $in: user.currentGroups } },
        { accessGroups: { $in: user.currentGroups } },
        { isPublished: true },
      ],
    };
    // ...
  }
  return datablockFilter;
}
```

**Analysis:** This method manually constructs MongoDB query filters. While it's functional, it could benefit from type safety improvements and better integration with the CASL ability system.

**Recommendation:** Consider using CASL's `accessibleBy` helper or creating a more type-safe filter builder. However, this is a pre-existing pattern in the codebase, not introduced by this refactoring.

## Verdict

### ✅ Strengths

1. **Architecture Improvement:** The extraction of datablock authorization logic into a dedicated class is a significant improvement. It follows SOLID principles and makes the code more maintainable.

2. **Simplified Action Enum:** Reducing 12 datablock-specific actions to 4 is a major simplification. The old naming (`DatablockCreateEndpoint`, `DatablockCreateInstance`, `DatablockCreateAny`) was confusing and inconsistent.

3. **Consistent Permission Logic:** The new implementation maintains the same security model as the old one. No permissions were accidentally loosened or tightened.

4. **Better Separation of Concerns:** The factory no longer needs to know about datablock-specific business rules.

5. **Reduced Code Duplication:** The old code had separate methods for endpoint access and instance access, which had overlapping logic. The new code unifies this.

### ⚠️ Issues Found

| Severity | Issue | File | Impact |
|----------|-------|------|--------|
| **HIGH** | `accessGroups` may be undefined, breaking all group-based permissions | `datablocks.ability.ts:22-24` | Users lose all group-based permissions |
| **MEDIUM** | Redundant permission declarations | `datablocks.ability.ts:79-83` | No functional impact, but confusing |
| **LOW** | Inconsistent error messages | `datablocks.controller.ts` | Minor debugging inconvenience |
| **INFO** | Delete permission doesn't check ownership | `datablocks.ability.ts:86-92` | Intentional (DELETE_GROUPS design) |

### 📊 Overall Rating: **7.5/10**

**Breakdown:**
- **Architecture:** 9/10 - Excellent refactoring, good separation of concerns
- **Security:** 8/10 - Maintains same security model, but one critical bug found
- **Code Quality:** 7/10 - Good overall, but has redundant code and minor inconsistencies
- **Testing:** 6/10 - No evidence of new tests for the refactored code

### 🎯 Final Recommendation

**DO NOT MERGE UNTIL CRITICAL BUG IS FIXED.**

The refactoring is well-designed and represents a significant improvement in code organization. However, the critical bug with `accessGroups` being potentially undefined must be fixed before merging.

**Required Fixes:**
1. Fix the `accessGroups` initialization to handle missing configuration (HIGH priority)

**Recommended Improvements:**
1. Remove redundant `can()` statements in `DatablockAbility.buildAbility()`
2. Standardize error messages in the controller
3. Add unit tests for `DatablockAbility` class
4. Document the DELETE_GROUPS unrestricted delete access design decision

**After Fixes:** The branch should be merged. The refactoring improves maintainability without compromising security (once the critical bug is fixed).

**Comment by nitrosx** [comment]
# Overview

## What the Code Does

The changes refactor datablock authorization in the CASL layer. The authorization logic controls datablock access based on:
- Public datablocks (`isPublished: true`) are readable by anyone
- Owners (`ownerGroup`) can read and update their datablocks
- Access group members (`accessGroups`) can read
- Privileged groups (`CREATE_DATASET*`) can create and update any datablock
- Admin groups (`ADMIN_GROUPS`) have full access
- Delete groups (`DELETE_GROUPS`) can delete any datablock

## Files Changed

| File | Change |
|------|--------|
| `src/casl/abilities/datablocks.ability.ts` | New: 100 lines, encapsulates datablock authorization |
| `src/casl/action.enum.ts` | Simplified: 12 old actions replaced with 4 new ones |
| `src/casl/casl-ability.factory.ts` | Refactored: removed 87 lines, delegates to new class |
| `src/casl/casl.module.ts` | Updated: registers `DatablockAbility` |
| `src/datablocks/datablocks.controller.ts` | Updated: uses new action names |

## Logic Correctness

The new implementation is functionally equivalent to master. The old code split authorization into `datablockEndpointAccess()` (endpoint-level) and `datablockInstanceAccess()` (instance-level). The new code unifies these into a single `buildAbility()` method.

All permission rules are preserved:
- Unauthenticated: read only published
- Authenticated: read owned, accessed, or published; update owned
- Privileged: create and update any
- Admin: full access
- Delete: delete any

## Edge Cases

### Handled
- Null user: only published read access
- Empty `currentGroups`: no special permissions
- Multiple groups: `some()` checks work correctly
- Missing `isPublished`: MongoDB handles gracefully

### Not Handled
- **Critical**: If `accessGroups` configuration is missing, `this.accessGroups` is undefined (line 23 of `datablocks.ability.ts`). All group-based permission checks silently fail. Admin, delete, and privileged group users lose all special permissions.

## Unreachable Code

No code blocks are completely unreachable. However, there are redundant declarations:
- Line 55: `can(Action.DatablockRead, Datablock, ifPublished)` - duplicate of line 38
- Lines 81-83: `can()` statements for create/read/update after `AccessAny` is granted - redundant but harmless

## Do the Changes Make Sense

Yes. The refactoring:
1. Improves separation of concerns by extracting datablock logic into its own class
2. Simplifies action naming from 12 actions to 4 clearer ones
3. Unifies endpoint and instance authorization, eliminating duplication
4. Follows established architecture patterns

## Summary

The refactoring maintains the same security model while improving code organization. The changes are sound and the architecture is better. One critical bug exists: missing `accessGroups` configuration will silently break group-based permissions. This must be fixed before merging.

**Verdict**: Architecturally sound. Fix `accessGroups` null handling, then merge.

**Comment by nitrosx** [comment]
# Security Review

## 1. Injection Vulnerabilities

**Status: None Found**

All MongoDB conditions use CASL's type-safe ability builder with values from validated JWT tokens. No dynamic query construction from user-supplied strings. No use of eval, $where, or other MongoDB injection vectors. Values used in conditions (`user.currentGroups`) are sourced from authenticated JWT, not raw user input.

## 2. Sensitive User Data Exposure

**Status: None Found**

No new endpoints introduced. No data serialization changes. Same data returned to clients. Only authorization logic refactored. Public/private data visibility controlled by `isPublished`, `ownerGroup`, `accessGroups` unchanged from master.

## 3. Insecure API Usage

**Status: None Found**

No new external API calls introduced. No hardcoded credentials, tokens, or secrets. No changes to configuration for external services. All changes are internal to the authorization layer.

## 4. Authentication Bypass

**Status: None Found - Security Improved**

The new implementation maintains the same security model as master and actually fixes one issue:

| Aspect | Master | New Code | Security Impact |
|--------|--------|----------|----------------|
| Unauthenticated read | Allowed for published datablocks | Allowed for published datablocks | Same |
| Owner read | ownerGroup in currentGroups | ownerGroup in currentGroups | Same |
| Access group read | accessGroups in currentGroups | accessGroups in currentGroups | Same |
| Owner update | **accessGroups in currentGroups** (BUG) | **ownerGroup in currentGroups** (FIXED) | **Security improved** |
| Privileged create | createDataset groups | createDataset groups | Same |
| Admin access | Full access | Full access via AccessAny | Same |
| Delete access | delete groups | delete groups | Same |

The old code incorrectly allowed datablock updates to any user in the accessGroups, not just owners. The new code correctly restricts updates to owners only.

## Critical Configuration Bug

**Issue**: In `DatablockAbility` constructor (line 22-23), `accessGroups` is retrieved from ConfigService without null handling:

```typescript
this.accessGroups = this.configService.get<AccessGroupsType>("accessGroups");
```

If the `accessGroups` configuration is missing, `this.accessGroups` will be undefined. All subsequent group-based permission checks using optional chaining (`this.accessGroups?.admin.includes(g)`) will silently fail, causing:
- Users in CREATE_DATASET groups to lose create/update permissions
- Users in ADMIN groups to lose admin permissions
- Users in DELETE groups to lose delete permissions

**Recommendation**: Use `getOrThrow()` or provide a default empty object to prevent silent failures.

## Verdict

**Overall: Secure with one critical configuration bug**

- No injection vulnerabilities introduced
- No sensitive data exposure
- No insecure API usage
- No authentication bypass possible; security actually improved by fixing owner update check
- One critical bug: missing accessGroups configuration silently breaks all group-based permissions

**Action Required**: Fix the accessGroups initialization to handle missing configuration before merging.

**Comment by bpedersen2** [comment]
The access groups remark from AI sounds reasonable.

And looking at the DELETE ability, are there real use cases where this is is not admin access? As delete applies to all datasets, I don't see any ...

**Comment by HayenNico** [comment]
@bpedersen2 I agree, datablocks in their current form actually have the most irregular permission model as compared to the other collections, with extra rights for a lot of groups.
Since this is supposed to be refactor only, I've reproduced the old behavior; fixing this is part of the redefinition/rework of permissions (discussion thread [here](https://github.com/SciCatProject/backend/discussions/2826)).

Regarding the accessGroups comment: I'll fix that for all casl PRs (it's the same everywhere, interestingly Sourcery only notices/complains about 50% of the time)

**Comment by nitrosx** [comment]
---

# Code Review Report: refactor-casl-factory-datablocks Branch

**Date:** 2026-08-13 13:31:00  
**Reviewer:** Mistral Vibe  
**Branch:** refactor-casl-factory-datablocks  
**Base Branch:** master  
**Commit Range:** 6c25dae3 (HEAD) - comparing against master

---

## Overview

### 1. Verify logic correctness

The refactoring introduces a new `DatablockAbility` class that centralizes all datablock authorization logic, replacing the previous scattered implementation across `datablockEndpointAccess` and `datablockInstanceAccess` methods in `CaslAbilityFactory`. The logic has been consolidated and simplified.

**Key logic changes:**
- **Unified authorization model**: Combines endpoint-level and instance-level authorization into a single `buildAbility` method
- **Simplified action enum**: Replaced 15 datablock-specific actions (`DatablockCreateEndpoint`, `DatablockReadEndpoint`, `DatablockUpdateEndpoint`, `DatablockDeleteEndpoint`, `DatablockCreateInstance`, `DatablockReadInstance`, `DatablockUpdateInstance`, `DatablockReadAny`, `DatablockUpdateAny`, `DatablockDeleteAny`) with 4 clear actions (`DatablockCreate`, `DatablockRead`, `DatablockUpdate`, `DatablockDelete`)
- **Cumulative permissions**: The new implementation properly chains permissions using CASL's logical OR, giving precedence to the broadest applicable rule

**Logic correctness assessment:** The refactored logic maintains the same authorization behavior while being more maintainable. The permission matrix is now explicitly documented and implemented consistently.

### 2. Check if all edge cases are handled

**Edge cases identified and handled:**

- **Unauthenticated users**: Can only read published datablocks (line 39, `datablocks.ability.ts`)
- **Authenticated users**: Can read if in ownerGroup, accessGroups, or if published (lines 54-56)
- **CREATE_DATASET_GROUPS users**: Can create and update any datablock (lines 60-74)
- **ADMIN_GROUPS users**: Have unrestricted access via `Action.AccessAny` (lines 76-85)
- **DELETE_GROUPS users**: Can read, update, and delete any datablock (lines 87-96)
- **Null user handling**: Proper early return for unauthenticated users (lines 41-46)

**Missing edge case handling:**
- No explicit handling for when `accessGroups` configuration is undefined (though defaulted to `{}` on line 24)
- No validation that `user.currentGroups` is not null/undefined for authenticated users

### 3. Summarize what the code touched by the changes does

The changes refactor the datablock authorization system by:

1. **Creating a new module** (`src/casl/abilities/datablocks.ability.ts`):
   - Encapsulates all datablock authorization logic in a dedicated, injectable service
   - Implements the permission model described in the new documentation

2. **Updating action enum** (`src/casl/action.enum.ts`):
   - Removed 15 redundant datablock actions
   - Added 4 simplified, consistent actions

3. **Refactoring factory** (`src/casl/casl-ability.factory.ts`):
   - Removed `datablockEndpointAccess` (34 lines) and `datablockInstanceAccess` (62 lines) methods
   - Added `datablockAccess` method that delegates to `DatablockAbility.buildAbility()`
   - Updated constructor to accept `DatablockAbility` dependency
   - Updated endpoint accessors to use new method name

4. **Updating module** (`src/casl/casl.module.ts`):
   - Added `DatablockAbility` to providers array

5. **Refactoring controller** (`src/datablocks/datablocks.controller.ts`):
   - Replaced all references from old actions to new simplified actions
   - Updated `checkPermission` method to use new `datablockAccess` method
   - Moved access control filter logic into controller as private method
   - Improved consistency in permission checking

6. **Adding documentation** (`docs/developer-guide/authorization/datablocks.md`):
   - Comprehensive permission matrix
   - Clear explanation of permission hierarchy
   - Implementation notes

### 4. Assess whether the changes make sense

**Yes, the changes make sense and represent a significant improvement:**

- **Reduced complexity**: 15 actions → 4 actions, eliminating confusion between endpoint vs instance actions
- **Improved maintainability**: Single source of truth for datablock authorization
- **Better separation of concerns**: Authorization logic moved to dedicated class
- **Enhanced consistency**: All datablock authorization follows the same pattern
- **Clear documentation**: Permission model is now explicitly documented

### 5. Identify any unreachable code

**No unreachable code identified** in the new implementation. The old code in master had potential issues:
- `datablockEndpointAccess` always returned the same permissions regardless of user state for some actions
- `datablockInstanceAccess` had duplicate and overlapping permission definitions

The refactored code eliminates these issues by using a unified approach.

---

## Code Changes

### Summary of Changes

| File | Change Type | Lines Added | Lines Removed | Net Change |
|------|-------------|--------------|---------------|------------|
| `docs/developer-guide/authorization/datablocks.md` | Added | 69 | 0 | +69 |
| `src/casl/abilities/datablocks.ability.ts` | Added | 103 | 0 | +103 |
| `src/casl/action.enum.ts` | Modified | 4 | 22 | -18 |
| `src/casl/casl-ability.factory.spec.ts` | Modified | 2 | 0 | +2 |
| `src/casl/casl-ability.factory.ts` | Modified | 5 | 155 | -150 |
| `src/casl/casl.module.ts` | Modified | 2 | 1 | +1 |
| `src/datablocks/datablocks.controller.ts` | Modified | 82 | 72 | +10 |
| **Total** | | **267** | **250** | **+17** |

### Assessment: Are the changes necessary?

**Yes, absolutely necessary.** The previous implementation had several issues:

1. **Action proliferation**: 15 different datablock actions created confusion and maintenance burden
2. **Duplicated logic**: Both endpoint and instance authorization methods contained overlapping permission logic
3. **Inconsistent patterns**: Different authorization approaches for different resources
4. **Poor separation of concerns**: Authorization logic was mixed into the factory class
5. **Lack of documentation**: No clear documentation of the permission model

### Improvement Needed

While the changes are largely positive, the following improvements are recommended:

1. **Add dedicated unit tests for `DatablockAbility`**: Currently, only the factory initialization is tested, not the actual permission logic (commit: `6c25dae3`)
2. **Handle null `accessGroups` more gracefully**: Consider throwing an error or logging a warning if configuration is missing
3. **Validate user object**: Add guard clauses to ensure `user.currentGroups` exists when user is not null
4. **Consider adding integration tests**: Test the full authorization flow from controller through factory to ability

### Verdict: Final judgment on the changes

**Verdict: APPROVED with minor recommendations**

The refactoring is well-executed and addresses significant technical debt. The changes:
- ✅ Maintain backward compatibility in behavior
- ✅ Improve code maintainability
- ✅ Follow best practices for separation of concerns
- ✅ Add comprehensive documentation
- ✅ Reduce complexity

The lack of dedicated unit tests for the new `DatablockAbility` class is the primary concern, but this does not block the changes.

---

## Security Review

### 1. Are there any potential injection vulnerabilities?

**No injection vulnerabilities identified.**

- The authorization logic uses CASL's MongoDB query builder, which properly escapes and validates conditions
- All database queries use parameterized queries through Mongoose
- No string concatenation or raw query building from user input

### 2. Does this code expose any sensitive user data?

**No sensitive data exposure identified.**

- The authorization logic only checks group memberships and datablock metadata
- User data is only accessed for `currentGroups` which is used for authorization decisions
- No personal data (emails, names, etc.) is exposed through the authorization system

### 3. Are there instances of insecure API usage?

**No insecure API usage identified.**

- All authorization checks use the standardized CASL pattern
- Permission checks are performed before any data access operations
- The `checkPermission` method properly throws `ForbiddenException` for unauthorized access

### 4. Could this code lead to an authentication bypass?

**No authentication bypass vulnerabilities identified.**

- The refactoring maintains the same security boundaries as the original code
- Unauthenticated users (null user) are explicitly handled with restricted permissions
- All authorization checks are performed consistently using the same ability object
- The new implementation actually *improves* security by:
  - Removing duplicate and potentially conflicting permission rules
  - Centralizing authorization logic for easier auditing
  - Using a more consistent permission model

---

## Test Coverage

### Current Test Coverage Assessment

**Inadequate for the new functionality.**

| Component | Test Coverage | Status |
|-----------|---------------|--------|
| `DatablockAbility` class | None | ❌ Missing |
| `datablockAccess` method in factory | Minimal (initialization only) | ⚠️ Incomplete |
| Controller authorization flow | None | ❌ Missing |
| Permission matrix validation | None | ❌ Missing |

### Test Coverage Details

**Existing tests:**
- `src/casl/casl-ability.factory.spec.ts`: Only tests factory initialization with `DatablockAbility`, not the actual authorization logic
- `src/datablocks/datablocks.service.spec.ts`: Tests service layer but not authorization

**Missing tests:**
1. No tests for `DatablockAbility.buildAbility()` with various user roles
2. No tests for the permission matrix (unauthenticated, authenticated, special groups)
3. No tests for edge cases (null user, null groups, empty groups)
4. No integration tests for the full authorization flow

### Suggestions for Improvement

1. **Create `datablocks.ability.spec.ts`**:
   ```typescript
   describe('DatablockAbility', () => {
     describe('Unauthenticated user', () => {
       it('should allow reading published datablocks');
       it('should deny creating datablocks');
       it('should deny updating datablocks');
       it('should deny deleting datablocks');
     });
     
     describe('Authenticated user', () => {
       it('should allow reading own datablocks');
       it('should allow reading access group datablocks');
       it('should allow reading published datablocks');
       it('should allow updating own datablocks');
       it('should deny updating others datablocks');
     });
     
     describe('CREATE_DATASET_GROUPS user', () => {
       it('should allow creating any datablock');
       it('should allow updating any datablock');
     });
     
     describe('ADMIN_GROUPS user', () => {
       it('should have unrestricted access');
     });
     
     describe('DELETE_GROUPS user', () => {
       it('should allow reading any datablock');
       it('should allow updating any datablock');
       it('should allow deleting any datablock');
     });
   });
   ```

2. **Add integration tests** in existing controller tests:
   - Test authorization in each controller method
   - Verify proper exception throwing for unauthorized access
   - Test access control filter application

3. **Add test for permission matrix** in documentation or as a truth table test

---

## Security Examples

*No security vulnerabilities were identified in the changes. This section remains empty as no relevant examples apply.*

---

## Testing for Security Use Cases

*No security vulnerabilities were identified. The following testing approach would validate the security of the authorization system:*

### Suggested Security Test Cases

1. **Unauthenticated Access Attempt**
   - **Test**: POST /datablocks without authentication
   - **Expected**: 403 Forbidden
   - **Code location**: `datablocks.controller.ts:90` - `checkPermission` with `Action.DatablockCreate`

2. **Cross-Group Data Access**
   - **Test**: GET /datablocks/:id where datablock belongs to different group
   - **Expected**: 403 Forbidden
   - **Code location**: `datablocks.controller.ts:234` - `checkPermission` with `Action.DatablockRead`

3. **Admin Bypass**
   - **Test**: DELETE /datablocks/:id with ADMIN_GROUPS user
   - **Expected**: 200 OK (or appropriate success response)
   - **Code location**: `datablocks.ability.ts:76-85` - Admin group permissions

4. **Elevated Privilege Validation**
   - **Test**: Verify CREATE_DATASET_GROUPS user can create datablock for any ownerGroup
   - **Expected**: 201 Created
   - **Code location**: `datablocks.ability.ts:60-74` - Create dataset group permissions

5. **Filter Injection Prevention**
   - **Test**: GET /datablocks with malformed filter parameter
   - **Expected**: Filter is properly sanitized or rejected
   - **Code location**: `datablocks.controller.ts:142` - `addAccessControlFilters`

---

## Summary

### Key Findings

1. **Refactoring Quality**: Excellent - The changes significantly improve code maintainability and consistency
2. **Security**: No vulnerabilities identified - Authorization logic is sound and follows best practices
3. **Functionality**: Backward compatible - Same behavior with cleaner implementation
4. **Documentation**: Comprehensive - New documentation clearly explains the permission model
5. **Test Coverage**: Inadequate - Missing dedicated tests for the new `DatablockAbility` class

### Recommendations

1. **Before merging**: Add unit tests for `DatablockAbility` class
2. **Optional**: Consider adding integration tests for the full authorization flow
3. **Future**: Apply the same refactoring pattern to other resource types (origdatablocks, proposals, etc.)

### Overall Assessment

**The refactoring represents a significant improvement in code quality and maintainability without introducing any security vulnerabilities or breaking changes. The changes are approved with the recommendation to add comprehensive unit tests for the new authorization logic.**

### Commits Responsible for Key Changes

- **Primary refactoring**: `d4b1f1da` - refactor datablocks casl & controller
- **Security fixes**: `6fce9bf1` - fix security gaps in datablocks controller  
- **Documentation**: `36d3655c` - add developer docs
- **Test updates**: `6c25dae3` - fix CaslAbilityFactory unit test initialization, lint

**Review by nitrosx — APPROVED** [review]
